### PR TITLE
Add an extension slot to the upload request.

### PR DESCRIPTION
**Pull request by chris-wood**
One possible extension would be an opaque blob to carry client attestation information (for #89), if desired. The current set is empty, but I suspect we'll need freedom to extend here as this gets used without totally changing the protocol version.

If folks are generally supportive of this, it should probably land after #90, which touches and renames PDAUploadReq to PDAReport.

### CODE REVIEW

File: draft-pda-protocol.md [hunk 1]
```diff
@@ -508,14 +508,20 @@ the leader's endpoint URL. The payload is structured as follows:
 struct {
   PDATaskID task_id;
   Time time;
+  Extension extensions<8..2^16-1>;
```
Comment (ekr) [inline draft-pda-protocol.md:511]: Isn't this actually 4..2^16-1

**Review by ekr — APPROVED** [review]

File: draft-pda-protocol.md (hunk 1 above)
Comment (chris-wood) [reply draft-pda-protocol.md:511]:
```suggestion
  Extension extensions<4..2^16-1>;
```

File: draft-pda-protocol.md [hunk 2]
```diff
@@ -508,14 +508,20 @@ the leader's endpoint URL. The payload is structured as follows:
...
   PDATaskID task_id;
   Time time;
+  Extension extensions<4..2^16-1>;
   PDAEncryptedInputShare encrypted_input_shares<1..2^16-1>;
 } PDAUploadReq;
 ~~~
 
+[OPEN ISSUE: if we include extensions in the AAD then these extensions are required to be
+echoed in each aggregate subrequest. We could either not include this in the AAD, or replace
+the list with a hash of the list (that's sent separately), or something else.]
```
Comment (cjpatton) [inline draft-pda-protocol.md:518]: In fact, when the extension is used for the purpose of attestation, it probably *needs* to be echoed in each aggregate sub-request.

File: draft-pda-protocol.md [hunk 3]
```diff
@@ -563,6 +569,27 @@ The leader responds to well-formed requests to `[leader]/upload` with status 200
 and an empty body. Malformed requests are handled as described in
 {{pa-error-common-aborts}}.
 
+### Upload Extensions {#upload-extensions}
+
+Each PAUploadReq carries a list of extensions that clients may use to convey
+additional, authenticated information in the report. Each extension is a tag-length
```
Comment (cjpatton) [inline draft-pda-protocol.md:575]:
```suggestion
additional, authenticated[TODO: Specify how] information in the report. Each extension is a tag-length
```

**Review by cjpatton — CHANGES_REQUESTED** [review]
I'm supportive of solving #89 in a generic way, but it's not clear that this change has enough plumbing. In particular, there doesn't appear to be a way for the helper to verify either property of the report (1 - the report was generated by a trusted client; 2 - the report was generated in a trusted environment) Minimally, I think we would need to echo the extension in the aggregate sub-request.

I'd be happy with one of of the following outcomes:
- Copy the extension into the aggregate sub-request and add an OPEN ISSUE for deciding whether this is sufficient plumbing for client attestation
- Close this PR and park the issue

File: draft-pda-protocol.md (hunk 3 above)
Comment (chris-wood) [reply draft-pda-protocol.md:575]: The contents of the extensions are presumably fed into the AAD, so closing.

File: draft-pda-protocol.md (hunk 2 above)
Comment (chris-wood) [reply draft-pda-protocol.md:518]: That is where I lean, too, but it does mean extension bloat in each request. That's probably fine, though.

**Comment by chris-wood** [comment]
Both (1) and (2) are _external to the protocol_. The purpose of extensions is to allow whatever mechanism exists for addressing them to be bound to the report, as is done here. 

> Copy the extension into the aggregate sub-request and add an OPEN ISSUE for deciding whether this is sufficient plumbing for client attestation

This is done in the issue. This is _not meant_ to close the attestation issue. It's meant to give us the mechanics to address it later.

**Review by cjpatton — APPROVED** [review]
Right you are. What threw me off was the following statement:
> [OPEN ISSUE: if we include extensions in the AAD then these extensions are required to be
> echoed in each aggregate subrequest. We could either not include this in the AAD, or replace
> the list with a hash of the list (that's sent separately), or something else.]

The "then these extensions are required to be echoed..." makes it sound as if they're *not* echoed, which they are. I'm happy with this PR modulo fixing this statement.

File: draft-pda-protocol.md [hunk 4]
```diff
@@ -703,6 +730,7 @@ unique client report. Sub-requests are structured as follows:
 ~~~
 struct {
   Time time; // Equal to PDAReport.time.
+  Extension extensions<8..2^16-1>;
```
Comment (tgeoghegan) [inline draft-pda-protocol.md:733]: Any reason this is `<8..2^16-1>` here but `<4..2^16-1>` in `PDAUploadReq`?

File: draft-pda-protocol.md [hunk 5]
```diff
@@ -1282,6 +1310,11 @@ the input shares and defeat privacy.
 
 # IANA Considerations
 
-TODO
+## Upload Extension Registry
+
+This document requests creation of a new registry for extensions to the Upload
+protocol. This registry should contain the following columns:
+
+[TODO: define how we want to structure this registry when the time comes]
```
Comment (tgeoghegan) [inline draft-pda-protocol.md:1318]: Do we need to provide this much structure for the extensions? IIUC at the PDA protocol layer, we will never examine or modify these extensions so why can't they be completely opaque blobs?

File: draft-pda-protocol.md (hunk 4 above)
Comment (chris-wood) [reply draft-pda-protocol.md:733]: Me not making the change in two places :)

Comment (chris-wood) [reply draft-pda-protocol.md:733]:
```suggestion
  Extension extensions<4..2^16-1>;
```

File: draft-pda-protocol.md (hunk 5 above)
Comment (chris-wood) [reply draft-pda-protocol.md:1318]: This just creates the registry where extensions will go. We don't need to do anything more here, unless we decide we want to introduce some mandatory extensions.

Comment (tgeoghegan) [reply draft-pda-protocol.md:1367]: Well, my point applies more to the `struct Extension` definition. The extension just be an opaque blob of bytes, and the responsibility of protocol implementations would be to relay the blob unmodified from `PDAReport` to `PDAAggregateSubReq`. Then the format of that blob is something to be agreed upon by the client that generated it and the aggregator that is examining it, out of band from the PDA protocol.

Comment (chris-wood) [reply draft-pda-protocol.md:1367]: The contents of the extension, and how they're used, are up to the thing that defines the extension. So, yeah, I think what we have here aligns with what you envision.

**Pull request merged**